src/lib/statistics/chunk_statistics/histograms/equal_element_count_histogram.cpp

```diff
@@ -0,0 +1,206 @@
+                                                                    const std::vector<std::string>& maxs,
+                                                                    const std::vector<HistogramCountType>& heights,
+                                                                    const HistogramCountType distinct_count_per_bin,
+                                                                    const BinID bin_count_with_extra_value,
+                                                                    const std::string& supported_characters,
+                                                                    const uint32_t string_prefix_length)
+    : AbstractHistogram<std::string>(supported_characters, string_prefix_length),
+      _mins(mins),
+      _maxs(maxs),
+      _heights(heights),
+      _distinct_count_per_bin(distinct_count_per_bin),
+      _bin_count_with_extra_value(bin_count_with_extra_value) {
+  DebugAssert(mins.size() > 0, "Cannot have histogram without any bins.");
+  DebugAssert(mins.size() == maxs.size(), "Must have the same number of lower as upper bin edges.");
+  DebugAssert(mins.size() == heights.size(), "Must have the same number of edges and heights.");
+  DebugAssert(distinct_count_per_bin > 0, "Cannot have bins with no distinct values.");
+  DebugAssert(bin_count_with_extra_value < mins.size(), "Cannot have more bins with extra value than bins.");
+
+  for (auto bin_id = 0u; bin_id < mins.size(); bin_id++) {
+    DebugAssert(heights[bin_id] > 0, "Cannot have empty bins.");
+    DebugAssert(mins[bin_id].find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+    DebugAssert(maxs[bin_id].find_first_not_of(supported_characters) == std::string::npos, "Unsupported characters.");
+    DebugAssert(mins[bin_id] <= maxs[bin_id], "Cannot have upper bin edge higher than lower bin edge.");
+
+    if (bin_id < maxs.size() - 1) {
+      DebugAssert(mins[bin_id] < mins[bin_id + 1], "Bins must be sorted and cannot overlap.");
+      DebugAssert(maxs[bin_id] < maxs[bin_id + 1], "Bins must be sorted and cannot overlap.");
+    }
+  }
+}
+
+template <typename T>
+EqualElementCountBinStats<T> EqualElementCountHistogram<T>::_get_bin_stats(
+    const std::vector<std::pair<T, HistogramCountType>>& value_counts, const BinID max_bin_count) {
+  // If there are fewer distinct values than the number of desired bins use that instead.
+  const auto bin_count = value_counts.size() < max_bin_count ? static_cast<BinID>(value_counts.size()) : max_bin_count;
+
+  // Split values evenly among bins.
+  const HistogramCountType distinct_count_per_bin = static_cast<HistogramCountType>(value_counts.size() / bin_count);
+  const BinID bin_count_with_extra_value = value_counts.size() % bin_count;
+
+  std::vector<T> mins;
+  std::vector<T> maxs;
+  std::vector<HistogramCountType> heights;
+  mins.reserve(bin_count);
```

Comment:
Doesn't your version first initialize the vector with default values of T, which are then all overwritten?
Less code, yes, but will it really be faster?

Reply:
We had a discussion about this somewhere. I think the crux was that for trivial data types - at least - reserve/push_back is slower because each push_back() call checks whether there is still capacity available.

Reply:
Okay, changed it in EqualWidth and EqualElementCount. For EqualHeight we don't necessarily know how many bins will actually be created, so the logic does not work there easily.